Improved patrol cat select & Random patrol groups

Treefire33 opened the pull request:
Changes the patrol cat select to instead be a list(?) of checkboxes with the sprite of each cat shown above. Also adds the ability to send a completely random group of cats on patrol.

The current implementation of the UI isn't final and I'm still trying to decide where exactly to place a selected cats info. Along with that I'm still not sure if displaying only 6 cats per page is the way to go about it.

https://github.com/user-attachments/assets/118d3973-659a-4d55-a968-cb86e4b07dd1

implements #25 
implements #24

larkgz commented:
Hey, I just pushed an update to the checkbox component that should allow you to make the CatDisplay a part of the label and therefore also clickable. Should result in slightly better UX and be a little easier to style custom checkboxes, if that's something you want to do.

I think my ideas for placing a selected cat's info would be:

<img width="219" height="269" alt="image" src="https://github.com/user-attachments/assets/6e2a0061-6a82-4af4-aa57-4af9e52175f2" />
(Possibly harder to style? Not sure though)

or
<img width="548" height="184" alt="image" src="https://github.com/user-attachments/assets/e0f95471-66c9-4a7b-8018-cfa90227afb3" />


Also, yes, I agree that 6 cats per page is maybe a little too restrictive, especially if there's no way of seeing what cats you have selected or filtering them.

One idea I had was to have there be an "add cat" button that opened a modal dialogue for cat selection with search, etc (and then the cat selection could maybe be reusable for other places like choosing a mate) but I don't know how that would feel in practice.

larkgz commented:
This is just feedback on the design so far, not the code: 
* Hmm, trying it out, I think I'd actually prefer if the random option just added a single random cat the way that the game did.
* I think the selected cats should be displayed somewhere on the page, because after you check a cat and change pages, they kind of vanish into the aether and it's hard to remember.
* Can the number of cats per page be upped to, like... at least 10? Maybe 18 or 24?
* Should probably be responsive using grid or flexbox (most likely grid).
* Cat sprites should also be clickable when selecting because it's more convenient. Can put React nodes into the `label` argument of the Checkbox so that the contents are all clickable (e.g. `label={<><CatDisplay cat={cat} w="75px" h="75px" /> {cat.name.display} </>}` should work)

Would be really nice but might be out of scope for this PR:
* Simple search by name
* A way to filter for mentors/apprentices of selected cats

larkgz commented:
A few last things, and then I think you can mark it for review: 
* Can the selected cats be displayed somewhere? Might have to be separate from the cat search component. 
* Can “select random group“ be changed to adding a single random cat?

Treefire33 commented:
I moved the cat selection buttons to the search element and added the select random cat button, while also keeping the select random all button. Along with that the currently selected patrol cats are displayed on the patrol page, which I might update to also include the information present in ClanGen's patrol menu (patrol skills and traits).

Last thing I want to add before I mark ready for review is extra filter categories, as only having the status filter select is very limiting. Though I don't know what categories would benefit the player (especially since I haven't played ClanGen in a while), so I wondering if you have any ideas as to what would be beneficial to the player?

larkgz commented on src/components/CatSearch.tsx:
```diff
@@ -0,0 +1,238 @@
+import { TbCaretLeftFilled, TbCaretRightFilled } from "react-icons/tb";
+
+import "../styles/cat-search.css";
+import "../styles/cats-page.css";
```
If you want to import these css styles, please put them into different file. `cats-page.css` should be reserved for css on the cats page.

larkgz commented on src/components/CatSearch.tsx:
```diff
@@ -0,0 +1,238 @@
+          <button tabIndex={0} 
+            onClick={() => selectRandom(maxSelection)}
+          >
+            Select Random All
```
```suggestion
            Select Random Group
```

larkgz commented on src/components/CatSearch.tsx:
```diff
@@ -0,0 +1,238 @@
+            <Checkbox label="Leader" checked={statusFilters["leader"]} onChange={() => toggleStatusFilter("leader")}/>
+          </ul>
+          <button tabIndex={0} onClick={() => toggleAllStatusFilters(true)}>Select All</button>
+          <button tabIndex={0} onClick={() => toggleAllStatusFilters(false)}>Deselect All</button>
```
```suggestion
          <div className="button-row">
            <button tabIndex={0} onClick={() => toggleAllStatusFilters(true)}>Select All</button>
            <button tabIndex={0} onClick={() => toggleAllStatusFilters(false)}>Deselect All</button>
          </div>
```

larkgz commented on src/components/CatSearch.tsx:
```diff
@@ -0,0 +1,238 @@
+  });
+
+  const [affiliationEnabled, setAffiliationEnabled] = useState<boolean>(false);
+  const [miscFilters, setMiscFilters] = useState<Record<string, any>>({
```
Why is the value `any` and not `boolean`?

larkgz commented on src/components/CatSearch.tsx:
```diff
@@ -0,0 +1,238 @@
+    }
+    if (miscFilters["apprentice"]) { // Valid if cat has an apprentice in selected cats.
+      let found = cat.apprentices.find(c => selectedCats.includes(c.ID));
+      metApprenticeFilter = found != undefined;
```
```suggestion
      metApprenticeFilter = found !== undefined;
```

larkgz commented on src/components/CatSearch.tsx:
```diff
@@ -0,0 +1,238 @@
+
+  function checkFilters(cat: Cat) {
+    let metNameFilter = searchName == "" ? true : cat.name.display.toLowerCase().includes(searchName); // True by default since name filter can be empty.
+    let metStatusFilter = statusFilters[cat.status.replace(/\s/g,'')];
```
Why not just keep the names in the filter object with spaces?

larkgz commented on src/components/CatSearch.tsx:
```diff
@@ -0,0 +1,247 @@
+import { TbCaretLeftFilled, TbCaretRightFilled } from "react-icons/tb";
+
+import "../styles/cat-search.css";
+import "../styles/cats-page.css";
```
For organizational purposes, I would prefer that `cats-page.css` only contains css exclusive to the `/cats` page. If using that css, please move it into a different file.

larkgz commented on src/components/CatSearch.tsx:
```diff
@@ -0,0 +1,247 @@
+          <button tabIndex={0} 
+            onClick={() => selectRandom(maxSelection)}
+          >
+            Select Random All
```
```suggestion
            Select Random Group
```
I think this is a little more clear.

larkgz commented on src/pages/PatrolsPage.tsx:
```diff
@@ -170,73 +143,31 @@ function PatrolsPage() {
+        />
+      </fieldset>
+
+      <fieldset className="cats-list">
```
Can this fieldset be the same height as the `CatDisplay` so it doesn't grow in height when cats are selected?

Treefire33 commented:
Most of the requested changes were implemented, and I also went ahead and tried to make the search element look better for smaller devices in portrait mode. I could try to also make the code shrink the amount of cats per page depending on the width but I think this works (though the cat icons may be a tad large).

https://github.com/user-attachments/assets/e6288dcb-0478-4c24-82ae-815d2760c90e
